libs/storage/Tsavorite/cs/src/core/Index/Checkpointing/StateMachineDriver.cs

```diff
@@ -23,28 +23,119 @@ public class StateMachineDriver
         SemaphoreSlim waitForTransitionIn;
         // All threads have exited the given state
         SemaphoreSlim waitForTransitionOut;
+        // Transactions drained in last version
+        public long lastVersion;
+        public SemaphoreSlim lastVersionTransactionsDone;
         List<IStateMachineCallback> callbacks;
         readonly LightEpoch epoch;
         readonly ILogger logger;
+        readonly long[] NumActiveTransactions;
 
         public SystemState SystemState => SystemState.Copy(ref systemState);
 
-        public StateMachineDriver(LightEpoch epoch, SystemState initialState, ILogger logger = null)
+        public StateMachineDriver(LightEpoch epoch, ILogger logger = null)
         {
             this.epoch = epoch;
-            this.systemState = initialState;
+            this.systemState = SystemState.Make(Phase.REST, 1);
             this.waitingList = [];
+            this.NumActiveTransactions = new long[2];
             this.logger = logger;
         }
 
         public void SetSystemState(SystemState state)
+            => systemState = SystemState.Copy(ref state);
+
+        internal long GetNumActiveTransactions(long txnVersion)
+            => Interlocked.Read(ref NumActiveTransactions[txnVersion & 0x1]);
+
+        void IncrementActiveTransactions(long txnVersion)
+            => _ = Interlocked.Increment(ref NumActiveTransactions[txnVersion & 0x1]);
+
+        void DecrementActiveTransactions(long txnVersion)
         {
-            systemState = SystemState.Copy(ref state);
+            if (Interlocked.Decrement(ref NumActiveTransactions[txnVersion & 0x1]) == 0)
+            {
+                if (lastVersionTransactionsDone != null && txnVersion == lastVersion)
+                {
+                    lastVersionTransactionsDone.Release();
+                }
+            }
         }
 
+        /// <summary>
+        /// Acquire a transaction version - this should be called before
+        /// BeginLockable is called for all sessions in the transaction.
+        /// </summary>
+        /// <returns></returns>
+        public long AcquireTransactionVersion()
+        {
+            var isProtected = epoch.ThisInstanceProtected();
+            if (!isProtected)
+                epoch.Resume();
+            try
+            {
+                // We create a barrier preventing new transactions from starting in the PREPARE_GROW phase
+                // since the lock table needs to be drained and transferred to the larger hash index.
+                while (systemState.Phase == Phase.PREPARE_GROW)
```

Comment:
Should this include Phase.PREPARE?

Reply:
PREPARE_GROW is specifically for index growth. That is the only case we need to barrier transactions in this PR.